File: src/tensor/add.cc
/*! @copyright (c) 2022-present Skolkovo Institute of Science and Technology
 *                              (Skoltech), Russia. All rights reserved.
 *                 2023-present Artificial Intelligence Research Institute
 *                              (AIRI), Russia. All rights reserved.
 *
 * NNTile is software framework for fast training of big neural networks on
 * distributed-memory heterogeneous systems based on StarPU runtime system.
 *
 * @file src/tensor/add.cc
 * Add operation for Tensor<T>'s
 *
 * @version 1.1.0
 * */

#include "nntile/tensor/add.hh"
#include "nntile/starpu/add.hh"

namespace nntile::tensor
{

//! Tensor-wise add operation
template<typename T>
void add_async(Scalar alpha, const Tensor<T> &src1, Scalar beta,
        const Tensor<T> &src2, const Tensor<T> &dst)
{
    // Check dimensions
    if(dst.ndim != src1.ndim)
    {
        throw std::runtime_error("dst.ndim != src1.ndim");
    }
    if(dst.ndim != src2.ndim)
    {
        throw std::runtime_error("dst.ndim != src2.ndim");
    }
    // Check shapes of tensors
    for(Index i = 0; i < dst.ndim; ++i)
    {
        if(dst.shape[i] != src1.shape[i])
        {
            throw std::runtime_error("dst.shape[i] != src1.shape[i]");
        }
        if(dst.shape[i] != src2.shape[i])
        {
            throw std::runtime_error("dst.shape[i] != src2.shape[i]");
        }
        if(dst.basetile_shape[i] != src1.basetile_shape[i])
        {
            throw std::runtime_error("dst.basetile_shape[i] != "
                    "src1.basetile_shape[i]");
        }
        if(dst.basetile_shape[i] != src2.basetile_shape[i])
        {
            throw std::runtime_error("dst.basetile_shape[i] != "
                    "src2.basetile_shape[i]");
        }
    }
    // Do nothing if alpha is zero
    if(alpha == 0.0 && beta == 1.)
    {
        return;
    }
    // Apply per-tile add asynchronously as needed
    int mpi_rank = starpu_mpi_world_rank();

    for(Index i = 0; i < src1.grid.nelems; ++i)
    {
        // Get handle for corresponding tiles of src and dst
        auto src1_tile_handle = src1.get_tile_handle(i);
        auto src2_tile_handle = src2.get_tile_handle(i);
        auto dst_tile_handle = dst.get_tile_handle(i);
        // MPI rank of the destination tile
        int dst_tile_rank = dst_tile_handle.mpi_get_rank();
        // Transfer data
        src1_tile_handle.mpi_transfer(dst_tile_rank, mpi_rank);
        src2_tile_handle.mpi_transfer(dst_tile_rank, mpi_rank);
        // Execute only on destination node
        if(mpi_rank == dst_tile_rank)
        {
            auto traits = src1.get_tile_traits(i);
            starpu::add::submit<T>(traits.nelems, alpha, src1_tile_handle,
                    beta, src2_tile_handle, dst_tile_handle);
        }
        // Flush cache for the output tile on every node
        dst_tile_handle.mpi_flush();
    }

}

//! Tensor-wise add operation
template<typename T>
void add(Scalar alpha, const Tensor<T> &src1, Scalar beta,
        const Tensor<T> &src2, const Tensor<T> &dst)
{
    add_async<T>(alpha, src1, beta, src2, dst);
    starpu_task_wait_for_all();
    starpu_mpi_wait_for_all(MPI_COMM_WORLD);
}

// Explicit instantiation of template
template
void add_async<fp32_t>(Scalar alpha, const Tensor<fp32_t> &src1, Scalar beta,
        const Tensor<fp32_t> &src2, const Tensor<fp32_t> &dst);

template
void add_async<bf16_t>(Scalar alpha, const Tensor<bf16_t> &src1, Scalar beta,
        const Tensor<bf16_t> &src2, const Tensor<bf16_t> &dst);

template
void add_async<fp32_fast_tf32_t>(Scalar alpha,
        const Tensor<fp32_fast_tf32_t> &src1, Scalar beta,
        const Tensor<fp32_fast_tf32_t> &src2,
        const Tensor<fp32_fast_tf32_t> &dst);

template
void add_async<fp64_t>(Scalar alpha, const Tensor<fp64_t> &src1, Scalar beta,
        const Tensor<fp64_t> &src2, const Tensor<fp64_t> &dst);

// Explicit instantiation of template

template
void add<fp32_t>(Scalar alpha, const Tensor<fp32_t> &src1, Scalar beta,
        const Tensor<fp32_t> &src2, const Tensor<fp32_t> &dst);

template
void add<bf16_t>(Scalar alpha, const Tensor<bf16_t> &src1, Scalar beta,
        const Tensor<bf16_t> &src2, const Tensor<bf16_t> &dst);

template
void add<fp32_fast_tf32_t>(Scalar alpha, const Tensor<fp32_fast_tf32_t> &src1,
        Scalar beta, const Tensor<fp32_fast_tf32_t> &src2,
        const Tensor<fp32_fast_tf32_t> &dst);

template
void add<fp64_t>(Scalar alpha, const Tensor<fp64_t> &src1, Scalar beta,
        const Tensor<fp64_t> &src2, const Tensor<fp64_t> &dst);

} // namespace nntile::tensor

File: wrappers/python/nntile/layer/add.py
# @copyright (c) 2022-present Skolkovo Institute of Science and Technology
#                              (Skoltech), Russia. All rights reserved.
#                2023-present Artificial Intelligence Research Institute
#                              (AIRI), Russia. All rights reserved.
#
# NNTile is software framework for fast training of big neural networks on
# distributed-memory heterogeneous systems based on StarPU runtime system.
#
# @file wrappers/python/nntile/layer/add.py
# Add layer of NNTile Python package.
# It is used in skip-connection operation
#
# @version 1.1.0

import nntile.utils.constructors as nntc
from nntile.layer.base_layer import BaseLayer
from nntile.tensor import (
    TensorMoments, TensorTraits, add_async, add_inplace_async)


class Add(BaseLayer):
    def __init__(self, x: TensorMoments, y: TensorMoments, res: TensorMoments):
        self.x = x
        self.y = y
        self.res = res
        super().__init__([x, y], [res], [], [])

    @staticmethod
    def generate_simple(x: TensorMoments, y: TensorMoments):
        res_traits = TensorTraits(y.value.shape, y.value.basetile_shape)
        # Create result tensor with the same distribution as y
        res_value = type(y.value)(res_traits, y.value.distribution)
        res_grad = type(y.value)(res_traits, y.value.distribution)
        res = TensorMoments(res_value, res_grad, True)
        return Add(x, y, res)

    def forward_async(self):
        add_async(1.0, self.x.value, 1.0, self.y.value, self.res.value)
        self.x.value.wont_use()
        self.y.value.wont_use()
        self.res.value.wont_use()

    def forward_dynamic(self, x1: TensorMoments, x2: TensorMoments):
        y = nntc.empty_like(x1.value)
        add_async(1.0, x1.value, 1.0, x2.value, y)
        return TensorMoments(y, None, False)

    def backward_async(self):
        add_inplace_async(1.0, self.res.grad, 1.0, self.x.grad)
        add_inplace_async(1.0, self.res.grad, 1.0, self.y.grad)
        self.x.grad.wont_use()
        self.y.grad.wont_use()
        self.res.grad.wont_use()
